Report progress of sfold operation

ttpcodes opened the pull request:
This PR modifies the logic of the `sfold` operation to allow for the `Progress` of the operation to be reported.

As part of the call to `FoldFile`, the user must provide a `struct` that implements `FoldFileOperations`, an interface declaring a `method` named `UpdateFoldFileProgress`, that will be used to report the status of the `FoldFile` operation via `callback`.

As part of the changes to allow for `sfold` to report its `Progress`, the `coalesce` function that `FoldFile` uses has been converted to a `goroutine`, meaning that `FoldFile` no longer blocks and will immediately return once the `sfold` operation has started. It is up to the user to implement `FoldFileOperations` and respond to the `Progress` changes of the operation accordingly.

The default implementations used by `sfold` in the `CLI` and in the tests have also been modified to implement `FoldFileOperations`, waiting for a `Done` status and sending any `error` that might've occurred during the operation.

yasker commented on sparse/sfold.go:
```diff
@@ -59,41 +62,67 @@ func coalesce(parentFileIo FileIoProcessor, childFileIo FileIoProcessor) error {
+	go func() {
+		var err error
+		var progress int
+		defer func() {
```
Use 
```
go func() (err error) {
}
```
instead so you won't need to remember to set error when return

yasker commented on sparse/sfold.go:
```diff
@@ -38,18 +43,16 @@ func FoldFile(childFileName, parentFileName string) error {
 	if err != nil {
 		panic("Failed to open childFile, error: " + err.Error())
 	}
-	defer childFileIo.Close()
```
If get blocksize or get exts errors out below, files won't be closed.

Also, need another commit to change all those `panic` into `return error`. Now sfold is in the main process, they cannot just panic.

yasker commented on sparse/sfold.go:
```diff
@@ -59,41 +62,67 @@ func coalesce(parentFileIo FileIoProcessor, childFileIo FileIoProcessor) error {
-				return err
+				ops.UpdateFoldFileProgress(progress, true, err)
+			} else {
+				ops.UpdateFoldFileProgress(100, true, nil)
```
Use constant instead of `100`

Commit:
Track progress of sfold operation via callback

Commit:
Convert panics in sfold to returned errors

ttpcodes commented:
- Set `error` in `coalesce` as named return parameter to avoid having to manually set `err`.
- Stored `100` as a `constant` value (`progressComplete`).
- Moved `DirectFileIoProcessor` creation into `coalesce` as part of `goroutine` to clean up `defer` logic.
- Replaced calls to `panic` with returns of `errors`.

yasker commented on sparse/sfold.go:
```diff
@@ -33,6 +37,22 @@ func FoldFile(childFileName, parentFileName string) error {
+
+	defer func() {
+		if err != nil {
+			ops.UpdateFoldFileProgress(progress, true, err)
```
Since previous we have logs for each statement, it's better we keep it that way to print log when an error happens as well.

yasker commented on sparse/sfold.go:
```diff
@@ -46,19 +66,20 @@ func FoldFile(childFileName, parentFileName string) error {
-		return err
+		return fmt.Errorf("failed to GetFiemapExtents of childFile filename: %s, err: %v", childFileIo.Name(), err)
 	}
+	childFileInfo, err := childFileIo.Stat()
```
Size can be passed in as a parameter since the caller has the information.

ttpcodes commented:
- Any `error` encountered during `coalesce` now triggers `log.Errorf` during the `defer` call.
- `fileSize` is now passed in as a parameter to avoid having to call `Stat` on the file twice.
